### ISSUE TITLE
[clsim] GPU detection broken for RTX-series cards (Trac #2212)

### ISSUE THREAD

**Issue by icecube-trac** [icecube-trac/tix2#2204]
Nvidia decided that ray-tracing was a thing, and renamed their GTX series "RTX", and e.g. the 2080 Ti advertises itself as "GeForce RTX 2080 Ti". clsim.traysegments.common.configureOpenCLDevices, however, special-cases cards named "GTX" to set the number of work items to something sane and enable native math. Is there any reason not to do away with the special cases, and simply scale the number of work items to be proportional to the global memory size, as well as enabling native math by default?

Migrated from https://code.icecube.wisc.edu/ticket/2212
```json
{
    "status": "closed", 
    "changetime": "2019-03-06T20:58:01", 
    "description": "Nvidia decided that ray-tracing was a thing, and renamed their GTX series \"RTX\", and e.g. the 2080 Ti advertises itself as \"GeForce RTX 2080 Ti\". clsim.traysegments.common.configureOpenCLDevices, however, special-cases cards named \"GTX\" to set the number of work items to something sane and enable native math. Is there any reason not to do away with the special cases, and simply scale the number of work items to be proportional to the global memory size, as well as enabling native math by default?", 
    "reporter": "jvansanten", 
    "cc": "", 
    "resolution": "fixed", 
    "_ts": "1551905881354849", 
    "component": "combo simulation", 
    "summary": "[clsim] GPU detection broken for RTX-series cards", 
    "priority": "normal", 
    "keywords": "", 
    "time": "2018-11-29T14:14:01", 
    "milestone": "", 
    "owner": "jvansanten", 
    "type": "defect"
}
```

**Comment by icecube-trac** [comment]
Trac update at 20181129T14:15:53: claudio.kopper commented:

> Turning off native math used to be necessary on very old (integrated/laptop) cards. I am pretty sure we have none of those left, so I don't think we need it.
> 
> So I am all for making this change!

**Comment by icecube-trac** [comment]
Trac update at 20181129T22:34:52: jvansanten commented:

> Capital. What was the original motivation for making the number of work items settable from the outside? Are there situations were this can't be calculated for the device, given the memory footprint of the kernel as compiled?

**Comment by icecube-trac** [comment]
Trac update at 20181129T22:47:52: claudio.kopper commented:

> I think this can always be calculated. If I remember correctly I wasn't sure at the time if this was something one should/could optimize so I made it configurable, but I do not think this is of much practical use.

**Comment by icecube-trac** [comment]
Trac update at 20181130T15:29:16: 
* jvansanten changed owner from "" to "jvansanten"
* jvansanten changed status from "new" to "closed"
* jvansanten changed resolution from "" to "fixed"
* jvansanten commented:

> In changeset 167193:
> 
> ```text
> #!CommitTicketReference repository="IceCube" revision="167193"
> Choose number of workitems for GPUs to be proportional to their memory
> footprint, rather than special-casing certain device names. Fixes #2204.
> ```

**Comment by icecube-trac** [comment]
Trac update at 20190112T02:40:10: nwhitehorn commented:

> In changeset 170462:
> 
> ```text
> #!CommitTicketReference repository="IceCube" revision="170462"
> Choose number of workitems for GPUs to be proportional to their memory
> footprint, rather than special-casing certain device names. Fixes #2204.
> ```

**Comment by icecube-trac** [comment]
Trac update at 20190306T20:43:02: 
* jvansanten changed status from "closed" to "reopened"
* jvansanten commented:

> changeset 167193 leads to sub-optimal configuration of certain older GTX series cards. For instance, on a GTX 980 pre-r167193, `common.configureOpenCLDevices()` returns
> 
> ```text
> [I3CLSimOpenCLDevice('NVIDIA CUDA', 'GeForce GTX 980', useNativeMath=True, approximateNumberOfWorkItems=1024000)]
> ```
> while post, we get
> 
> ```text
> [I3CLSimOpenCLDevice('NVIDIA CUDA', 'GeForce GTX 980', useNativeMath=False, approximateNumberOfWorkItems=1024)]
> ```
> and approximately 2.4x slower benchmark performance (52.8 Mph/s -> 22.0 Mph/s). `approximateNumberOfWorkItems` is ignored, but useNativeMath=False causes the kernel to use different math instructions. Are there any GPUs for which the `native_` variants of math built-ins are unsafe?


* jvansanten changed resolution from "fixed" to ""

**Comment by icecube-trac** [comment]
Trac update at 20190306T20:46:43: claudio.kopper commented:

> > Are there any GPUs for which the `native_` variants of math built-ins are unsafe?
> 
> This used to be necessary for old driver versions when I first wrote clsim. I don't think it is necessary anymore and we just might want to remove it (and always use "native math" on GPUs. In the past there were also some CPU-based OpenCL drivers that caused issues when "nativeMath" was turned on, so we might want to leave it off for CPUs? Or do more testing for CPUs.
> 
>

**Comment by icecube-trac** [comment]
Trac update at 20190306T20:58:01: 
* jvansanten changed status from "reopened" to "closed"
* jvansanten commented:

> In changeset 171812:
> 
> ```text
> #!CommitTicketReference repository="IceCube" revision="171812"
> Always use native math on GPUs. Fixes #2204 (again)
> ```


* jvansanten changed resolution from "" to "fixed"
